### PR TITLE
Ruby Refactored To New Code Model

### PR DESCRIPTION
**Pull request by fearthecowboy**

### CODE REVIEW

File: src/core/AutoRest/AutoRest.json [hunk 1]
```diff
@@ -6,6 +6,12 @@
     "Azure.CSharp": {
       "type": "PluginCsa, AutoRest.CSharp.Azure"
     },
+        "Ruby": {
```
Comment (vishrutshah) [inline src/core/AutoRest/AutoRest.json]: minor: indentations

Comment (fearthecowboy) [inline src/core/AutoRest/AutoRest.json]: Yeah, I saw that too. That's what I get for working on three branches concurrently that all affect the same file.

I'll make sure those are cleaned up before the promotion to `master`

Comment (vishrutshah) [inline src/core/AutoRest/AutoRest.json]: 👍

File: src/generator/AutoRest.Ruby.Azure/CodeGeneratorRba.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                if ((model.Extensions.ContainsKey(AzureExtensions.ExternalExtension) &&
+                    (bool)model.Extensions[AzureExtensions.ExternalExtension])
+                    || model.Name == "Resource" || model.Name == "SubResource")
+                {
+                    continue;
+                }
+
+                if( codeModel.pageModels.Any( each => each.Name.EqualsIgnoreCase(model.Name ) ) )
+                {
+                    Console.WriteLine($"[HEY!] Skipping Model:{model.Name} for PageModel");
```
Comment (vishrutshah) [inline src/generator/AutoRest.Ruby.Azure/CodeGeneratorRba.cs]: >HEY [](start = 41, length = 3)

Curious why are we skipping ?

File: src/generator/AutoRest.Ruby/Model/RequirementsRb.cs [hunk 3]
```diff
@@ -184,11 +152,11 @@ public virtual string GetDependencyGems()
 require 'concurrent'
 require 'ms_rest'";
 
-            if(!string.IsNullOrWhiteSpace(this.ns))
+            if(!string.IsNullOrWhiteSpace(Settings.Instance.Namespace))
             {
                 requirements = requirements 
                     + Environment.NewLine 
-                    + string.Format(CultureInfo.InvariantCulture, "require '{0}/{1}/module_definition'", GeneratedFolderName, this.sdkName);
+                    + string.Format(CultureInfo.InvariantCulture, "require '{0}/{1}/module_definition'", GeneratedFolderName, GeneratorSettingsRb.Instance.sdkName);
```
Comment (vishrutshah) [inline src/generator/AutoRest.Ruby/Model/RequirementsRb.cs]: >GeneratorSettingsRb.Instance.sdkName) [](start = 126, length = 37)

I am noticing different behavior for this value after generation. 

It supposed to generate a line like `require 'generated/azure_mgmt_authorization/module_definition'` but it's not generating line as `require 'generated/authorization/module_definition'`

Did you notice that?

File: src/generator/AutoRest.Ruby.Azure/CodeGeneratorRba.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            // Paged Models
+            foreach (var pageModel in codeModel.pageModels)
+            {
+                var pageTemplate = new PageModelTemplate { Model = pageModel };
+                await Write(pageTemplate, Path.Combine(GeneratorSettingsRb.Instance.modelsPath, CodeNamer.UnderscoreCase(pageModel.Name) + ImplementationFileExtension));
+            }
+
+            // Enums
+            foreach (var enumType in codeModel.EnumTypes)
+            {
```
Comment (vishrutshah) [inline src/generator/AutoRest.Ruby.Azure/CodeGeneratorRba.cs]: There is something changes here...for the enum model name it should 

```
    module PolicyType
      NotSpecified = "NotSpecified"
      BuiltIn = "BuiltIn"
      Custom = "Custom"
    end
```
https://github.com/Azure/azure-sdk-for-ruby/blob/master/management/azure_mgmt_policy/lib/generated/azure_mgmt_policy/models/policy_type.rb#L11

but now it is 

```
    module policyType
      NotSpecified = "NotSpecified"
      BuiltIn = "BuiltIn"
      Custom = "Custom"
    end
```

capitalization on module name is different.

Comment (fearthecowboy) [inline src/generator/AutoRest.Ruby.Azure/CodeGeneratorRba.cs]: Hmm. Lemme look into it.

Comment (fearthecowboy) [inline src/generator/AutoRest.Ruby.Azure/CodeGeneratorRba.cs]: Ok, I PascalCase'd the module name for enums.

Comment (fearthecowboy) [inline src/generator/AutoRest.Ruby.Azure/CodeGeneratorRba.cs]: Can you pull and regenerate the SDK and validate?

File: src/generator/AutoRest.Ruby/Templates/EnumTemplate.cshtml [hunk 5]
```diff
@@ -8,7 +8,7 @@ module @Settings.Namespace
     #
     @WrapComment("# ", "Defines values for " + Model.Name)
     #
-    module @(Model.Name)
+    module @(CodeNamer.Instance.PascalCase(Model.Name))
```
Comment (vishrutshah) [inline src/generator/AutoRest.Ruby/Templates/EnumTemplate.cshtml]: Ohhh i though fix should be inside then the template as the same names are being used in requirement templates as well :(

**Pull request merged**